Libplanet/Blockchain/BlockChain.Recalculate.cs

```diff
@@ -0,0 +1,165 @@
+        /// blocks in <see cref="IStateStore"/> are somehow corrupted.
+        /// </para>
+        /// </remarks>
+        internal void ComplementAllBlockStates(BlockHash blockHash)
+        {
+            _logger.Verbose("Complementing all block states upto {BlockHash}...", blockHash);
+
+            // Prevent recursive trial to recalculate & complement incomplete block states by
+            // mistake; if the below code works as intended, these state completers must never
+            // be invoked.
+            StateCompleterSet<T> stateCompleters = StateCompleterSet<T>.Reject;
+
+            // Calculates and fills the incomplete states on the fly.
+            foreach (BlockHash hash in BlockHashes)
+            {
+                Block<T> block = this[hash];
+                if (StateStore.ContainsBlockStates(hash))
+                {
+                    continue;
+                }
+
+                IReadOnlyList<ActionEvaluation> evaluations = ActionEvaluator.Evaluate(
+                    block,
+                    stateCompleters);
+
+                _rwlock.EnterWriteLock();
+                try
+                {
+                    SetStates(block, evaluations);
+                }
+                finally
+                {
+                    _rwlock.ExitWriteLock();
+                }
+
+                if (blockHash.Equals(hash))
+                {
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Recalculates and complements all <i>missing</i> block states including and upto given
+        /// <paramref name="blockHash"/> starting from the last known states in
+        /// <see cref="StateStore"/> before <paramref name="blockHash"/> if the states are missing
+        /// for <paramref name="blockHash"/>.
+        /// </summary>
+        /// <param name="blockHash">The starting point for searching backwards.</param>
+        internal void ComplementLatestBlockStates(BlockHash blockHash)
+        {
+            _logger.Verbose("Complementing latest block states upto {BlockHash}...", blockHash);
+
+            // Prevent recursive trial to recalculate & complement incomplete block states by
+            // mistake; if the below code works as intended, these state completers must never
+            // be invoked.
+            StateCompleterSet<T> stateCompleters = StateCompleterSet<T>.Reject;
+            Stack<BlockHash> stack = new Stack<BlockHash>();
+            BlockHash? pointer = blockHash;
+
+            while (pointer is { } p)
+            {
+                if (StateStore.ContainsBlockStates(p))
+                {
+                    break;
+                }
+                else
+                {
+                    stack.Push(p);
+                    pointer = this[p].PreviousHash;
+                }
+            }
+
+            foreach (BlockHash hash in stack)
```

Comment:
It looks like `ComplementAllBlockStates()` recalculates from older blocks to newer blocks, but this stack stored in inverse way. It is okay with that?

Reply:
When iterating over `Stack<T>`, iteration is done in popping order (without actually popping), i.e. how stacks would usually be consumed.